File: dataset/ZeroShot/CLAUDE_claude-sonnet-4-5-20250929/Task96.py

import socket
import threading

DEFAULT_PORT = 8080
BUFFER_SIZE = 1024

def handle_client(client_socket, address):
    """Handle individual client connection"""
    try:
        print(f"Client connected: {address}")
        while True:
            message = client_socket.recv(BUFFER_SIZE).decode('utf-8')
            if not message:
                break
            print(f"Received: {message}")
            client_socket.send(message.encode('utf-8'))  # Echo back
    except Exception as e:
        print(f"Client handler error: {e}")
    finally:
        client_socket.close()
        print(f"Client disconnected: {address}")

def start_server(port):
    """Start the echo server on specified port"""
    server_socket = None
    try:
        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server_socket.bind(('0.0.0.0', port))
        server_socket.listen(5)
        print(f"Server started on port {port}")
        
        while True:
            client_socket, address = server_socket.accept()
            client_thread = threading.Thread(target=handle_client, args=(client_socket, address))
            client_thread.daemon = True
            client_thread.start()
    except Exception as e:
        print(f"Server error: {e}")
    finally:
        if server_socket:
            server_socket.close()

if __name__ == "__main__":
    print("Echo Server Test Cases")
    print("======================")
    print("Test Case 1: Start server on default port 8080")
    print("Test Case 2: Server handles multiple clients")
    print("Test Case 3: Server echoes messages correctly")
    print("Test Case 4: Server handles client disconnection")
    print("Test Case 5: Server runs continuously")
    print("\\nStarting server...")
    
    start_server(DEFAULT_PORT)

File: dataset/ZeroShot/CLAUDE_claude-sonnet-4-5-20250929/Task96.cpp

#include <iostream>
#include <string>
#include <cstring>
#include <thread>
#include <vector>

#ifdef _WIN32
    #include <winsock2.h>
    #include <ws2tcpip.h>
    #pragma comment(lib, "ws2_32.lib")
    typedef int socklen_t;
#else
    #include <sys/socket.h>
    #include <netinet/in.h>
    #include <arpa/inet.h>
    #include <unistd.h>
    #define SOCKET int
    #define INVALID_SOCKET -1
    #define SOCKET_ERROR -1
    #define closesocket close
#endif

const int DEFAULT_PORT = 8080;
const int BUFFER_SIZE = 1024;

void handleClient(SOCKET clientSocket) {
    char buffer[BUFFER_SIZE];
    int bytesReceived;
    
    while ((bytesReceived = recv(clientSocket, buffer, BUFFER_SIZE - 1, 0)) > 0) {
        buffer[bytesReceived] = '\\0';
        std::cout << "Received: " << buffer << std::endl;
        send(clientSocket, buffer, bytesReceived, 0); // Echo back
    }
    
    closesocket(clientSocket);
    std::cout << "Client disconnected" << std::endl;
}

void startServer(int port) {
#ifdef _WIN32
    WSADATA wsaData;
    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        std::cerr << "WSAStartup failed" << std::endl;
        return;
    }
#endif

    SOCKET serverSocket = socket(AF_INET, SOCK_STREAM, 0);
    if (serverSocket == INVALID_SOCKET) {
        std::cerr << "Socket creation failed" << std::endl;
#ifdef _WIN32
        WSACleanup();
#endif
        return;
    }
    
    int opt = 1;
    setsockopt(serverSocket, SOL_SOCKET, SO_REUSEADDR, (char*)&opt, sizeof(opt));
    
    sockaddr_in serverAddr;
    memset(&serverAddr, 0, sizeof(serverAddr));
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_addr.s_addr = INADDR_ANY;
    serverAddr.sin_port = htons(port);
    
    if (bind(serverSocket, (sockaddr*)&serverAddr, sizeof(serverAddr)) == SOCKET_ERROR) {
        std::cerr << "Bind failed" << std::endl;
        closesocket(serverSocket);
#ifdef _WIN32
        WSACleanup();
#endif
        return;
    }
    
    if (listen(serverSocket, 5) == SOCKET_ERROR) {
        std::cerr << "Listen failed" << std::endl;
        closesocket(serverSocket);
#ifdef _WIN32
        WSACleanup();
#endif
        return;
    }
    
    std::cout << "Server started on port " << port << std::endl;
    
    while (true) {
        sockaddr_in clientAddr;
        socklen_t clientLen = sizeof(clientAddr);
        SOCKET clientSocket = accept(serverSocket, (sockaddr*)&clientAddr, &clientLen);
        
        if (clientSocket == INVALID_SOCKET) {
            std::cerr << "Accept failed" << std::endl;
            continue;
        }
        
        std::cout << "Client connected" << std::endl;
        std::thread(handleClient, clientSocket).detach();
    }
    
    closesocket(serverSocket);
#ifdef _WIN32
    WSACleanup();
#endif
}

int main() {
    std::cout << "Echo Server Test Cases" << std::endl;
    std::cout << "======================" << std::endl;
    std::cout << "Test Case 1: Start server on default port 8080" << std::endl;
    std::cout << "Test Case 2: Server handles multiple clients" << std::endl;
    std::cout << "Test Case 3: Server echoes messages correctly" << std::endl;
    std::cout << "Test Case 4: Server handles client disconnection" << std::endl;
    std::cout << "Test Case 5: Server runs continuously" << std::endl;
    std::cout << "\\nStarting server..." << std::endl;
    
    startServer(DEFAULT_PORT);
    
    return 0;
}
